packages/runtime-common/realm-auth-data-source.ts

```diff
@@ -0,0 +1,105 @@
+import { MatrixClient } from './matrix-client';
+import { RealmAuthClient } from './realm-auth-client';
+import { Loader } from './loader';
+
+type RealmInfo = {
+  isPublicReadable: boolean;
+  url: string;
+};
+
+export interface IRealmAuthDataSource {
+  get originalRealmURL(): string | undefined;
+  getRealmInfo(url: string): Promise<RealmInfo | null>;
+  getJWT(realmURL: string): Promise<string>;
+  resetAuth(realmURL: string): void;
+}
+
+type RealmInfoAndAuth = RealmInfo & { realmAuthClient?: RealmAuthClient };
+
+export class RealmAuthDataSource implements IRealmAuthDataSource {
+  // Cached realm info and session to avoid fetching it multiple times for the same realm
+  private visitedRealms = new Map<string, RealmInfoAndAuth>();
+  private matrixClient: MatrixClient;
+  private loader: Loader;
+  private _originRealmURL: string;
+
+  constructor(matrixClient: MatrixClient, loader: Loader, realmURL: string) {
+    this.matrixClient = matrixClient;
+    this.loader = loader;
+    this._originRealmURL = realmURL;
+  }
+
+  get originalRealmURL(): string {
+    return this._originRealmURL;
```

Comment:
Why not simply have a readonly property `realmURL`? I don't see why we'd have to name this differently and use a getter to access it.